In [2]:
from qiskit import QuantumCircuit
from src.quantum_noise import add_quantum_noise, analyze_noise_effect
from src.gate_basis_transformer import GateBasisTransformer
from src.drapper import test_quantum_adder
from src.quantum_noise_analyse import comprehensive_noise_analysis, demonstrate_noise_analysis

### Task 2.1

In [9]:
# Create a simple test circuit
qc = QuantumCircuit(2)
qc.h(0)    # Single-qubit gate
qc.cx(0,1) # Two-qubit gate
qc.h(1)    # Single-qubit gate

# Add noise
p1 = 0.01  # 1% error rate for single-qubit gates
p2 = 0.02  # 2% error rate for two-qubit gates
noisy_circuit = add_quantum_noise(qc, p1, p2)

# Analyze noise effects
stats = analyze_noise_effect(qc, p1, p2, num_samples=1000)
print(stats)

{'avg_x_errors': 0.018, 'avg_y_errors': 0.019, 'avg_z_errors': 0.017, 'avg_single_qubit_errors': 0.054, 'avg_two_qubit_errors': 0.0}


### Task 2.2

In [10]:
# Create a test circuit with non-basis gates
qc = QuantumCircuit(2)
qc.h(0)      # Hadamard gate (not in basis)
qc.cx(0, 1)  # CX gate (in basis)
qc.t(1)      # T gate (not in basis)
qc.y(0)      # Y gate (not in basis)

# Create transformer and transform circuit
transformer = GateBasisTransformer()
transformed_circuit = transformer.transform_circuit(qc)

# Get gate counts in the transformed circuit
gate_counts = transformer.get_gate_counts(transformed_circuit)
print("Gates in transformed circuit:", gate_counts)

Gates in transformed circuit: {'x': 0, 'rz': 9, 'cx': 1, 'sx': 6, 'id': 0}


### Test 2.3

In [11]:
# Example usage
a = 3
b = 5
test_results = test_quantum_adder(a, b, shots=1000)

print(f"Adding {a} and {b}:")
print(f"Expected sum: {test_results['expected_sum']}")
print(f"Number of qubits used: {test_results['n_qubits_used']}")
print("Measurement results:", test_results['measurement_counts'])

Adding 3 and 5:
Expected sum: 8
Number of qubits used: 4
Measurement results: {'1000': 1024}


### Task 2.4

In [3]:
# Basic analysis
a, b = 3, 5
results = comprehensive_noise_analysis(a, b, shots=1000)

# Detailed demonstration
analysis_results = demonstrate_noise_analysis()

Analyzing quantum addition of 3 + 5

Noise Level Analysis:
Noise levels (p1=0.001, p2=0.002): Success rate = 0.00%
Noise levels (p1=0.01, p2=0.02): Success rate = 0.00%
Noise levels (p1=0.05, p2=0.1): Success rate = 0.00%
Noise levels (p1=0.1, p2=0.2): Success rate = 0.00%

Circuit Depth Impact:
Circuit depth: 152
Total gates: 278
Success rate: 0.00%

Mitigation Recommendations:
Circuit optimization needed: True
Maximum recommended noise levels: {'max_single_qubit_error': 0.0, 'max_two_qubit_error': 0.0}
Reliability threshold: 0.00%


**Noise Level Analysis:**

At all the tested noise levels (p1=0.001, p2=0.002; p1=0.01, p2=0.02; p1=0.05, p2=0.1; p1=0.1, p2=0.2), the success rate for the quantum addition of 3 + 5 is 0.00%.
This indicates that the quantum circuit is extremely sensitive to noise, and even relatively low noise levels completely destroy the accuracy of the addition operation.


**Circuit Depth Impact:**

The circuit depth is 152, and the total number of gates is 278.
Despite these relatively high values, the success rate is still 0.00%, even with the lowest noise levels tested.


**Mitigation Recommendations:**

The analysis recommends that circuit optimization is needed, as the current circuit is not robust enough to handle any level of noise.
The suggested maximum noise levels for both single-qubit (p1) and two-qubit (p2) errors is 0.0, which means the circuit cannot tolerate any noise at all.
The reliability threshold is also 0.00%, further emphasizing the fragility of the current quantum addition circuit.



These results indicate a major problem with the quantum addition circuit in the presence of noise. The circuit is simply not resilient enough to handle even small amounts of noise, leading to a complete failure of the addition operation.